### ISSUE TITLE
FBuildWorker is not doing much

### ISSUE THREAD

**Issue by twallrab** [fastbuild/fastbuild#696]
My expectations were that connected FBuildWorkers would run compile jobs on all cores (mode:Work for others always). But surprisingly only a small part of the computing time (5%-10%) is used on the FBuildWorker machines, while the host computer is running at 100%. Mostly only 1 or 2 cpus (of 12 cpus) are used at the same time. While i am watching the output log, i noticed that a remote compile job is only spawned when a local job has just finished, instead of starting new jobs when the remote machine is idle. The network connection is fast enough. I can't figure out what the problem is. Maybe someone had the same problem and has a solution?

**Comment by claimred** [comment]
I've noticed the same, seems like cores spend a lot of time idling for some reason on certain projects. When a new target starts building it evenly distributes it's jobs but then distribution becomes uneven and workers are getting at most a tiny job. 

![image](https://user-images.githubusercontent.com/7941502/74665249-5f1c3700-51b0-11ea-9c96-c337b17b5287.png)

**Comment by ffulin** [comment]
*TL;DR:*

FASTBuild is often bound on local preparation of jobs (preprocessor speed). If you are not already, make sure you are using [Unity](https://fastbuild.org/docs/functions/unity.html). This makes everything (including local compilation and linking) as much as 10x faster.

The local host will ultimately always be the bottleneck of course.

*Details;*

There are a few things that impact how effective a remote compilation can be.

1) *Available workers.* The workers must be available and ready to accept jobs. This is the most obvious part, and if your workers are set to "Always" it looks like you have this covered.

2) *Local host bottleneck.* When in distributed mode, the local host will prioritize preparing jobs and handing them out. If the local machine is not fast enough to keep the remote workers fed, then they can be idle. This is the most common cause of worker under-utilization. You can usually see if this is the case by looking at the output when running from the command line:
```C++
20% [****-----] (x/y) + (z/w)
```
where:
 - x = active jobs running locally
 - y = queued local-only jobs
 - z = active remote jobs
 - w = queued remote jobs

If 'z' and 'w' are equal (or close to equal) it indicates that all possible remote work is being done remotely. (Note that in the absence of local work, local cpus can also do remote work)

In the case of a bottleneck like this, it can be improved by:
 - Having a better host machines (of course :) )
 - Using unity (this significantly reduces local processing, remote compilation and network transfer overheads - as much as 10x)
 - Using a modern compiler (if you're stuck on an old version of VS in particular (2015 or earlier) the preprocessor is quite slow

3) *Network bandwidth.* If the network is saturated, remote machines can be unable to receive work. Looking at the local host's network bandwidth use is an indicated, but if you have a slow (or congested) network, you'll need to investigate that with your IT department.

4) *Blocker dependencies in your build graph.* Some tasks can only be performed locally. If those tasks are dependencies of other remote work, that remote work cannot be prepared for distribution. Normal code doesn't have this, but you can check that you don't have unnecessary PreBuildDependencies which would cause artificial bottlenecks.  The -showdeps command line option can be used to emit the dependency tree, albeit in a rather unwieldy format.

**Comment by claimred** [comment]
@ffulin Thanks for in-depth explanation! Looks like local host bottleneck is exactly my case. I tried switching host machines and some of them utilize agents much, much better. Not sure what's the root cause of it though, CPU power looks roughly the same.

**Comment by claimred** [comment]
Speaking of slow preprocessor, is it possible to use clang-cl preprocessor using custom `.Preprocessor` and still compile code using old cl.exe? Would it be of any benefit?

**Comment by ffulin** [comment]
**> is it possible to use clang-cl preprocessor using custom .Preprocessor and still compile code using old cl.exe?**

Yes. Depending on which flavor of clang you are using, if you need different options for the preprocessor, you might need to additionally set .PreprocessorOptions

**> Would it be of any benefit?**

It depends. If you are CPU bound and the specific versions of the MSVC compiler and Clang compiler in question. It used to be (around the VS2013 era) that Clang's preprocessor was significantly faster than MSVCs, but these days there is much less of a difference.

Are you utilizing Unity? If not, that almost certainly give a significantly larger benefit than a preprocessor switch as it usually reduces the preprocessing overhead by an order of magnitude. (In addition to speeding up all the other parts of the build)

**Comment by john-blackburn** [comment]
I get the impression fbuild does not like having lots of workers in distributed mode. I just did an experiment with 3 workers (using the explict `.Workers` list) and got about 50% utilisation on each remote machine. I then used just 1 remote worker and got near 100% utilisation. I suspect (but can't yet prove) that increasing the number of workers further will just reduce CPU utilisation on each remote proportionately. 

So I think throwing more worker machines only adds marginally to the build speed and could even be counterproductive. A better solution is to have a worker list of 1 (using `.Workers` with a fixed IP hardwired) and that should be the most powerful computer you've got dedicated completely to compile help (eg with 32-64 cores). I haven't tested yet but I suspect that would give red hot performance.

**Comment by claimred** [comment]
I got a feeling that it severely depends on the nature of the codebase.

- For one project we've got enormous speedup gains from distributed compilation alone just by throwing more workers at it. Linearly shrinked from 45m to ~2:30m. All 15 workers are 4 cores middleclass intel cpus. 

It's a static library, no boost, fairly light C++ without dependencies, precompiled headers and templates, a lot of medium sized files (basically "handmade" unity style).

- For another project distributed compilation made building slower by 30% at first. We got a few more workers going and managed to get to the break-even point. It's not exactly practical though. Gonna use cache and unity for that beast.

Static library, precompiled header, boost, templates, a lot of small .cpp files which include precompiled header.

It'd be cool to try a monster machine with a lot of cores. I wonder if it is better to have several powerful workers or a lot of mediocre ones? One thing I noticed is that warm .fdb files seem to make workers utilization better. I think @ffulin mentioned somewhere that fdbs store timing data among other things.

> Are you utilizing Unity? 

Not at the moment, definitely going to try it later on. Not that trivial to fix name conflicts.

**Comment by john-blackburn** [comment]
@claimred Well I tried it out with a big 24 core machine and indeed got much better results than 5, 4-core machines. But the most crucial factor is network speed. I feel certain the graph you presented above would improve with network speed. This discussion is very significant:

https://github.com/fastbuild/fastbuild/issues/248

You can see 1 Gbps is essential. And both local and remote machines need a fixed ethernet cable, not WiFi!

**Comment by ffulin** [comment]
A gigabit network connection or better is almost certainly required to achieve any sort of reasonable level of scaling. Wi-fi of any kind is awful for anything throughput or latency sensitive (and FASTBuild distribution is both).

> One thing I noticed is that warm .fdb files seem to make workers utilization better. I think @ffulin mentioned somewhere that fdbs store timing data among other things.

The .fbd does store timing info and when it's available, it's used to prioritize the order of jobs. This should prioritize the higher cost items first, and exists specifically to improve overall subsequent build times (whether distributed or not)

> precompiled header, boost, templates, a lot of small .cpp .... distributed compilation made building slower by 30% at first. 

This is probably one of the worst case examples to try and improve compile times for :)

I probably don't have to say much about boost - it's impact on compile times I feel is well known. Its use is banned in every code base I am involved in. I imagine if you could, you'd probably have gotten rid of it already.

Lots of small cpp files as noted previously and in other threads, is the least efficient way to compile code. Unity is the best way to improve every aspect of compilation and linking (local, distributed or cached).

Precompiled headers can't be used on remote machines for distributed compilation, because of how distributed compilation is implemented in FASTBuild. This means they are effectively disabled for tasks built remotely. This lifts the break-even point a little bit, which is usually ok, but when combined with individual loose cpp files (and boost) this can add a signifcant overhead. This can also result in a lot of link-time overhead. Once again, Unity is the primary solution for this (both the compile and link-time overhead). Assuming you are using MSVC, the latest versions of Visual Studio have made very large gains in link-time performance which can help offset some of the increased link time overhead that can sometimes be seen (as well as just being faster in general).

**Comment by claimred** [comment]
>  ..the latest versions of Visual Studio have made very large gains in link-time performance which can help offset some of the increased link time overhead that can sometimes be seen 

Yeah, using -dist made our object files significantly bigger. Combined with /Z7 it made linking times long as hell. But! We also got huge link time improvement by using modern lld-link.exe instead of v140 link.exe, without /opt:ifc & /opt:ref. Link time hit is compensated in the end.

**Comment by ffulin** [comment]
I have heard people getting good speedups with lld-link as you've said and was considering trying that, but didn't bother when VS2019 Update 2 came along and improved link times by ~4x. If you switched to lld-link prior to that specific version, it might even be worth re-evaluating. It's great to have options :)

Interestingly, in the past, I've seen /opt:icf and /opt:ref slow down some links and speed up others, sometimes in different ways depending on how they are combined. I recall a situation where individually they were speedups, but when combined were slower! Again, the balance may have shifted with the most recent version(s).

**Comment by ffulin** [comment]
There isn't anything specifically actionable on this issue, and it's also quite old, so I'm going to close it.

There have been several versions of FASTBuild released since this issue was raised with several optimizations to compression, worker discovery, the network protocol and both local and remote scheduling. As of v1.04 there is also a -profile command line option that can be used to generate an overview of build scheduling and a -dot command line that can be used to help understand build bottlenecks.

Naturally we can re-open or create new issues as needed.
